### ISSUE TITLE
Do not run `no-unused-imports` rule by default

### ISSUE THREAD

**Issue by paul-dingemans** [pinterest/ktlint#3038]
The `no-unused-imports` rules tries to detect imports that are not used. Over the years, it has been proven difficult for KtLint to prevent false positive detections of unused imports. When removing an import as it is considered unused, while in fact it is used, this results in code that no longer can compile. This forces users to suppress the violation case-by-case, or to enable the rule in the project. The default behavior of KtLint should not force to do this.

Ktlint analyses code on a file by file base. Meaning, that all linting/formatting decisions have to be made based on information in the file that is being processed. Information in other files is not taken into account at all. For a 100% reliable detection whether an import is used, or unused, a file needs to be analyzed within the scope of the project and its dependencies.

The `no-unused-imports` rule follows several heuristic rules to mark an interface as used or unused with information in the file being processed. As no information outside of the context of the file is looked up, this results in false positives (an import marked as unused, while it is effectively required for compilation), and false negatives (an import is not marked as unsed, while it effectively is unused).

Given all problems with this problem (see list below), the rule will no longer be run by default. 
* #3022 
* #2940 
* #2855 
* #2821 
* #2670
* #2415 
* #2343 
* #1915
* #1792
* #1754
* #1587
* #1393
* #1367
* #1325
* #1277
* #1256
* #1050
* #1040
* #996
* #902
* #836
* #709
* #405
* #356

The `no-unused-imports` rule is marked as deprecated starting from Ktlint `1.7.0`, and will be removed in Ktlint 2.0.0. Until the rule is removed, it can be used by explicitly enabling the rule in the `.editorconfig`, by adding:
```
ktlint_standard_no-unused-imports = enabled # must be explicitly enabled as otherwise it won't run
```

**Comment by pkubowicz** [comment]
I understand it may be impossible to get 100% reliable detection of _all kinds_ of unused imports (wildcard, class, function imports etc.). However, I see a value from detecting _some kind_ of unused imports.

I think IntelliJ fails to mark a class as unused if it's referenced in import only. Which means unused imports can make it much harder to remove unused code, and as result can make it harder to reduce project complexity.

I would love to see a new rule that works on a limited category of imports (non-wildcard, non-function? or the opposite, function-only?), but works without false positives. Maybe there is a way to cover 20% of cases that bring 80% of value. Because every correctly detected unused import is a small step to better code maintenance.

**Comment by paul-dingemans** [comment]
> I understand it may be impossible to get 100% reliable detection of _all kinds_ of unused imports (wildcard, class, function imports etc.). However, I see a value from detecting _some kind_ of unused imports.

I am not worried about not detecting some imports that are actually unused. But removing imports that are used is really hurting users, as their code no longer compiles.

> I think IntelliJ fails to mark a class as unused if it's referenced in import only. Which means unused imports can make it much harder to remove unused code, and as result can make it harder to reduce project complexity.

I would be really surprised by that. The static analyzers which can take the full project into scope, compared to Ktlint which only takes a single file in scope, are way more powerfull in analyzing issues like this.
 
> I would love to see a new rule that works on a limited category of imports (non-wildcard, non-function? or the opposite, function-only?), but works without false positives. Maybe there is a way to cover 20% of cases that bring 80% of value. Because every correctly detected unused import is a small step to better code maintenance.

I don't think that such a rule belongs in Ktlint. But you can always develop such a rule, and open source it in a [custom ruleset](https://pinterest.github.io/ktlint/latest/api/custom-rule-set/) for everybody to use.

**Comment by rynkowsg** [comment]
I've been relying this rule for years now, and I was caught by surprise to see it disabled.

I worked with the rule only in Android projects, but I have never experienced false positives.

It would be a shame to remove this rule completely. Seems to me like throwing the baby out with the bathwater.

**Comment by eygraber** [comment]
Why remove the rule? Is there a specific issue with leaving it as opt-in?

I've been using it for many years with no issue, and it has been very valuable to me.

**Comment by patrickfeltes** [comment]
+1 also have not experienced false positives in Android and it is quite important in our project :( would love to keep this as opt-in.

**Comment by paul-dingemans** [comment]
The rule will be kept as opt-in in 2.x. release. Support on future shortcomings of this rule are not guaranteed to be fixed by the maintainers of this project.
